### DISCUSSION TITLE
Question About A .Net MAUI App Solution Structure

### DISCUSSION THREAD

**Discussion by MaroisConsulting** [dotnet/maui#7911] [General]
I'm working on a Project Management constuction app. It has a WPF Desktop project and a Xamarin Forms tablet project In the same solution.

The WPF app has views specific to the desktop app, and the Xamarin Forms project has views that are specific to the tablet. But both projects shared a lot of common things, like View Models, Data Models, Enums, Services, API proxy, etc. 

For example, the Xamarin Login View is entirely different then the Desktop Login View. Yet the common LoginViewModel works for both. The Desktop app has a Project View, but the Xamarin Forms project has a small subset in a smaller view. They share a VM also.

The point here is that the Desktop parts are in a WPF project, and the mobile parts are in an Xamarin Forms project. 

With .Net MAUI, there is only one project. It seems to me you can't possibly write a "shared" view that would work in both targets. And even if you could, you would almost certainly need #IF/#ELSE scattered all over to decide what view to show depending on what target you're running on. 

So, what I'm looking for here is a clearer understanding of how a .Net MAUI app is structured.

Thanks

**Comment by Quack-marsky** [comment]
I am no contributor nor member of MAUI team, but  from what I have seen, programmers already encountered this problem before. 
One solution is to go with web route, and provide views that would adjust themselves to the size app requires. 
Given that MAUI is not fit for such behavior (except Blazor apps maybe), I think the approach that programmers take working with Flutter will be more fit - they tend to separate views for types of devices (with separate directories for example), namely for desktop and mobile, and then displaying what is needed, keeping data sources shared. This in turn means that either you end up with unused views in the bundle or you have to put in conditional compilation to save space. Neither of them is ideal, but it makes it at least to have structure, and with that it will be easier to work with on a daily basis.

**Reply by MaroisConsulting** [reply]
@MyLastWish I agree with everything you've said. The approach of having different view for different platforms does make sense from an organizational perspective, but I agree that you would end up with views not used in different targets. There has to be a better approach.

**Comment by xtuzy** [comment]
MAUI solve difference platform use like MainPage.iOS.cs, MainPage.Windows.cs, see https://github.com/dotnet/maui/tree/main/src/Core/src/Handlers/Border , it define a https://github.com/dotnet/maui/blob/main/.nuspec/Microsoft.Maui.Controls.MultiTargeting.targets used in .csproj.
Use this, your code can not use ugly #if#else😊

**Reply by Quack-marsky** [reply]
Well, that looks promising

**Reply by MaroisConsulting** [reply]
Well that's just on page - the main page. But in an app targeting Desktop and Mobile, there are undoubtedly going to be dozens or even hundreds of view. Unless I'm totally missing somethine here, they are still all in the same solution.

If you were trying to share View Models, which is good design, you would still need #if/#endif code in the VM to handle platform specific code.
